### DISCUSSION TITLE
Define a custom Type to handle async file storage

### DISCUSSION THREAD

**Discussion by aminalaee** [sqlalchemy/sqlalchemy#10020] [Usage Questions]
Hey,

I'm trying to define a custom type called `FileType` using `TypeDecorator`.
The purpose is to save the file to local/remote storage when the instance is being saved and I have already done it in `process_bind_param`. Not sure if that's the best way to do it.

This works fine but trying to use it with asyncio I think this blocking call to save will block all other tasks, so I wanted to know what is the proper way of doing this with async/await?
I also checked events but I think they also don't have an async API.

```py
class FileType(TypeDecorator):
    impl = Unicode
    cache_ok = True

    def __init__(self, *args: Any, **kwargs: Any) -> None:
        super().__init__(*args, **kwargs)

    def process_bind_param(self, value: Any, dialect: Dialect) -> Optional[str]:
        # save file
        return value.filename

    def process_result_value(
        self, value: Any, dialect: Dialect
    ) -> bytes
        # read the file
```

Any help would be appreciated.

**Comment by zzzeek** [comment]
a few ways, easiest would be that your FileType returns a readable buffer object which will consume from the file when actually accessed, this would remove the overhead of reading out the file for each column in each row.   we would assume these files are small in any case so in practice I wouldnt actually worry about trying to make these work in non-blocking fashion - asyncio is totally serialized in any case so reading a file with blocking IO only means the task preemption spends a little more time on these objects.

to actually make these files read using non-blocking IO is possible using semi-private SQLAlchemy APIs but I wouldnt think is really worth it; these files are either very small, or you wouldn't want to consume them as result rows are read in any case.

**Reply by aminalaee** [reply]
Hey @zzzeek ,
Thanks a lot for the answer. But if I understand correctly you are describing the `process_result_value`, but my problem is storing the file in `process_bind_param`. I want to save the file when the instance is inserted.
Right now I'm doing it sync, both for sync and async, my last solution is to do something like `model.file.upload()` to upload the file, unless SQLAlchemy has a way that I am missing.

**Reply by zzzeek** [reply]
> but my problem is storing the file in process_bind_param. I want to save the file when the instance is inserted.

oh!  that's easy.  use a threadpool and use fire-and-forget.  unless you want the statement to raise an error if the save fails.

**Reply by zzzeek** [reply]
im not familiar with "file.upload()"

**Reply by aminalaee** [reply]
That's cool, yes "file.upload()" was just an example method to upload the file :smile: 
Thank you for the reply and help.
